text/3430-unix-socket-ancillary-data-v2.md

```diff
@@ -0,0 +1,896 @@
+
+fn recv_file(stream: &UnixStream) -> std::io::Result<File> {
+    // TODO: expose CMSG_SPACE() in a user-friendly way.
+    const ANCILLARY_CAPACITY: usize = 100;
+
+    let mut ancillary_buf = AncillaryDataBuf::with_capacity(ANCILLARY_CAPACITY);
+    let mut ancillary = ancillary_buf.to_ancillary_data();
+
+    let mut buf = [0u8; 1];
+    MessageReceiver::new(stream, &mut buf)
+        .ancillary_data(&mut ancillary)
+        .recv()?;
+
+    let mut received_fds: Vec<_> = ancillary.received_fds().collect();
+    if received_fds.len() != 1 {
+        // TODO: error handling (std::io::Error if not enough FDs returned)
+        panic!("didn't receive enough FDs");
+    }
+    let received_fd = received_fds.pop().unwrap();
+    Ok(File::from(received_fd))
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Control messages
+
+Each control message is represented as a `ControlMessage` struct, which holds
+a reference to the underlying data.
+
+```rust
+struct ControlMessage<'a>;
+
+impl ControlMessage<'a> {
+    fn new(
+        cmsg_level: c_int,
+        cmsg_type: c_int,
+        data: &'a [u8],
+    ) -> ControlMessage<'a>;
+
+    // Opaque platform-specific integers; same as `struct cmsghdr` fields.
+    fn cmsg_level(&self) -> c_int;
+    fn cmsg_type(&self) -> c_int;
+
+    // The type-specific data of this control message.
+    fn data(&self) -> &[u8];
+
+    // Whether this control message is truncated, such as by being received
+    // into a too-short buffer.
+    fn truncated(&self) -> bool;
+}
+```
+
+A `&ControlMessages` represents a buffer full of control messages. It can be
+inspected and iterated over to obtain `ControlMessage` values.
+
+
+```rust
+struct ControlMessages;
+
+impl ControlMessages {
+    fn from_bytes(bytes: &[u8]) -> &ControlMessages;
+    fn as_bytes(&self) -> &[u8];
+    fn is_empty(&self) -> bool;
+    fn iter(&self) -> ControlMessagesIter<'_>;
+}
+
+impl<'a> IntoIterator for &'a ControlMessages {
+    type Item = ControlMessage<'a>;
+    type IntoIter = ControlMessagesIter<'a>;
+}
+
+struct ControlMessagesIter<'a>;
+
+impl<'a> Iterator for ControlMessagesIter<'a> {
+    type Item = ControlMessage<'a>;
+}
+
+impl ControlMessagesIter<'a> {
+    // For inspecting non-iterable fragment in truncated buffer.
+    fn into_bytes(self) -> &'a [u8];
+}
+```
+
+## Ancillary data
+
+### `struct AncillaryData`
+
+An `AncillaryData` is responsible for combining the serialized control messages
+with a notion of file descriptor ownership, ensuring that (1) borrowed FDs live
+long enough to be sent, and (2) received FDs aren't leaked.
+
+```rust
+struct AncillaryData<'a, 'fd>;
+
+impl Drop for AncillaryData;
+
+impl AncillaryData<'a, 'fd> {
+    fn new(
+        control_messages_buf: &'a mut [MaybeUninit<u8>],
+    ) -> AncillaryData<'a, 'fd>;
+
+    // returns initialized portion of `control_messages_buf`.
+    fn control_messages(&self) -> &ControlMessages;
```

Comment:
The current implementation of `unix_socket_ancillary_data` (in nightly) does use an enum, but as noted in the RFC's first section that enum has turned out to be one of the biggest problems.

The Rust standard library can't have enums with cases that only exist on some platforms, and there's only one control message type specified by POSIX -- `SCM_RIGHTS`. So you end up with an enum that looks something like this:

```rust
pub enum ScmRights<'a> {
	Borrowed(&'a [BorrowedFd<'a>]),
	Owned(Vec<OwnedFd>),  // heap allocation!
}

pub struct GenericControlMessage<'a> { ... }

pub enum ControlMessage<'a> {
	ScmRights(ScmRights<'a>),
	Other(GenericControlMessage<'a>),
}
```

Given that the `SCM_RIGHTS` is also the only control message type that represents unmanaged resources transferred across the socket boundary, it makes sense to treat it specially inside the library. That leaves `enum ControlMessage` with only a single case, at which point it can be collapsed.

> Maybe there should be an associated type for the socket to advertise what type of control messages are supported. Then the iterator could use that to interpret messages into something more useful.

The maintenance load of cataloging `SCM_*` types across OSes would be very high, and documentation is thin regarding which control messages are supported by any given socket type. I don't think there's any way we can expect the Rust standard library to provide types for more than a small handful of the most popular ones.

Reply:
Hmm, I get your point. If the enum is *only* used for reporting control messages from the kernel, then it wouldn't be too bad if it contains variants that don't exist on all platforms. They would simply never be constructed.

But I do agree that cataloguing *all* messages is too much, considering all the platform specific differences.

So it probably does make sense to implement a higher level API for end-users on top of a pretty low level API. That could certainly be left for crates.io libraries though.